### PR TITLE
[flutter_tools] Introduce the new `flutter create-plugin` command.

### LINKED ISSUE #70873: Implement the new `flutter create plugin` command.
The `flutter create plugin` command will be a substitution of the old `flutter create -t plugin` command. 

This new command should have mostly the same usage as the old one expect:
1. The `--platforms` flag is enforced.
2. The `--sample` flag is not supported (this was only supported for flutter create -t app`
3. The `--list-samples` flag is not supported as it doesn't affect the the `-t plugin` route.

### PR DESCRIPTION
**Pull request by cyanglaz**
## Description

This PR introduces a new `flutter create-plugin` command.
**NOTE**: the original discussion is to add a subcommand for `flutter create`. We can't do that because `flutter create` still needs to work. So I ended up making it a separate command as `flutter create-plugin`. 
For now, this new `flutter create-plugin` can be run on the same directory to add platforms. After landing this, I'll create another PR to disable this functionality and add a `flutter add-platforms` command. See https://github.com/flutter/flutter/issues/59493

CC @amirh 

This PR didn't add a deprecation message for the `flutter create -t plugin` yet, I'll create a separate PR to do that.

## Related Issues

Fixes: https://github.com/flutter/flutter/issues/70873

## Tests

I added the following tests:

create_plugin_test.dart

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [x] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [x] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [x] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [x] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter_tools/lib/src/commands/create_plugin.dart [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    templateContext['androidPluginIdentifier'] = androidPluginIdentifier;
+
+    generatedCount += await generateApp(project.example.directory, templateContext, overwrite: overwrite, pluginExampleApp: true);
+    return generatedCount;
+  }
+
+  void _validateRequestedPlatforms(List<String> platforms) {
+    if (platforms == null || platforms.isEmpty) {
+      throwToolExit(_kNoPlatformsMessage,
+        exitCode: 2);
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/create_plugin.dart:154]:
```suggestion
      throwToolExit(_kNoPlatformsMessage, exitCode: 2);
```

File: packages/flutter_tools/lib/src/commands/create_plugin.dart [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    if (disabledPlatforms.isEmpty) {
+      return;
+    }
+    throwToolExit('''
+Requested platforms: $disabledPlatforms were disabled.
+For more information on desktop platforms, see https://flutter.dev/desktop#set-up.
+For more information on web, see https://flutter.dev/docs/get-started/web#set-up.
+
+''', exitCode: 2);
+
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/create_plugin.dart:178]:
```suggestion
```

File: packages/flutter_tools/lib/src/commands/create_plugin.dart [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    if (platforms.contains('macos') && !featureFlags.isMacOSEnabled) {
+      disabledPlatforms.add('macos');
+    }
+    if (platforms.contains('windows') && !featureFlags.isWindowsEnabled) {
+      disabledPlatforms.add('windows');
+    }
+    if (disabledPlatforms.isEmpty) {
+      return;
+    }
+    throwToolExit('''
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/create_plugin.dart:172]: this is confusing - we're default to include web and desktop platforms even though for most users they are not available?

File: packages/flutter_tools/test/commands.shard/permeable/create_plugin_test.dart [hunk 4]
```diff
@@ -0,0 +1,1225 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/commands.shard/permeable/create_plugin_test.dart:1]: How much of this code is just copy-pasted from the create test?

File: packages/flutter_tools/test/commands.shard/permeable/create_plugin_test.dart [hunk 5]
```diff
@@ -0,0 +1,1225 @@
...
+
+import '../../src/common.dart';
+import '../../src/context.dart';
+import '../../src/pubspec_schema.dart';
+import '../../src/testbed.dart';
+
+const String _kNoPlatformsMessage = 'Must specify at least one platform using --platforms.';
+const String frameworkRevision = '12345678';
+const String frameworkChannel = 'omega';
+// TODO(fujino): replace FakePlatform.fromPlatform() with FakePlatform()
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/commands.shard/permeable/create_plugin_test.dart:38]: please fix the todos if you copy-paste code

File: packages/flutter_tools/lib/src/commands/create_base.dart [hunk 6]
```diff
@@ -478,6 +498,36 @@ abstract class CreateBase extends FlutterCommand {
     return segments.join('.');
   }
 
+  /// Prints the directory information of generated plugins.
+  void printPluginDirectoryLocationMessage(String pluginPath, String projectName, String platformsString) {
+    final String relativePluginMain = globals.fs.path.join(pluginPath, 'lib', '$projectName.dart');
+    final String relativeExampleMain = globals.fs.path.join(pluginPath, 'example', 'lib', 'main.dart');
+    globals.printStatus('''
+
+    Your plugin code is in $relativePluginMain.
```
Comment (xster) [inline packages/flutter_tools/lib/src/commands/create_base.dart:507]: You sure the indents are intended?

File: packages/flutter_tools/lib/src/commands/create_plugin.dart [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+
+  @override
+  String get invocation => '${runner.executableName} $name <output directory>';
+
+  @override
+  Future<Map<CustomDimensions, String>> get usageValues async {
+    return <CustomDimensions, String>{
+      CustomDimensions.commandCreateAndroidLanguage: stringArg('android-language'),
+      CustomDimensions.commandCreateIosLanguage: stringArg('ios-language'),
+    };
```
Comment (xster) [inline packages/flutter_tools/lib/src/commands/create_plugin.dart:40]: don't you want some analytics for which platforms were plugins created for?

**Comment by cyanglaz** [comment]
Talked to @csells offline and he doesn't like the `flutter create-plugin` name. I'm going to close this PR for now and we'll discuss this further offline.

**Pull request closed without merging**